cmd/snap-update-ns/utils.go

```diff
@@ -132,105 +133,89 @@ func (sec *Secure) OpenPath(path string) (int, error) {
 	return fd, nil
 }
 
-// MkPrefix creates directories for all but the last segment and returns the
-// file descriptor to the leaf directory. This function is a base for secure
-// variants of mkdir, touch and symlink. None of the traversed directories
-// can be symbolic links.
-func (sec *Secure) MkPrefix(segments []string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) (int, error) {
-	logger.Debugf("secure-mk-prefix %q %v %d %d -> ...", segments, perm, uid, gid)
-
-	// Declare var and don't assign-declare below to ensure we don't swallow
-	// any errors by mistake.
-	var err error
-	var fd int
+// MkPrefix creates all the missing directories in a given base path and
+//
+// returns the file descriptor to the leaf directory. This function is a base
+// for secure variants of mkdir, touch and symlink. None of the traversed
+// directories can be symbolic links.
+func (sec *Secure) MkPrefix(base string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) (int, error) {
+	iter, err := NewPathIterator(base)
+	if err != nil {
+		// TODO: remove this and adjust tests
+		return -1, fmt.Errorf("cannot split unclean path %q", base)
+	}
+	iter.Next()
+	if iter.CurrentName() != "/" {
+		return -1, fmt.Errorf("path %v is not absolute", base)
+	}
 
 	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
-
 	// Open the root directory and start there.
-	fd, err = sysOpen("/", openFlags, 0)
+	fd, err := sysOpen("/", openFlags, 0)
 	if err != nil {
 		return -1, fmt.Errorf("cannot open root directory: %v", err)
 	}
-	if len(segments) > 1 {
+	for iter.Next() {
+		// Keep closing the previous descriptor as we go, so that we have the
+		// last one handy from the MkDir below.
 		defer sysClose(fd)
-	}
-
-	if len(segments) > 0 {
-		// Process all but the last segment.
-		for i := range segments[:len(segments)-1] {
-			fd, err = sec.MkDir(fd, segments, i, perm, uid, gid)
-			if err != nil {
-				return -1, err
-			}
-			// Keep the final FD open (caller needs to close it).
-			if i < len(segments)-2 {
-				defer sysClose(fd)
-			}
+		fd, err = sec.MkDir(fd, iter.CurrentBase(), iter.CurrentCleanName(), perm, uid, gid)
+		if err != nil {
+			return -1, err
 		}
 	}
 
-	logger.Debugf("secure-mk-prefix %q %v %d %d -> %d", segments, perm, uid, gid, fd)
 	return fd, nil
 }
 
-// MkDir creates a directory at segNum-th entry of absolute path represented
-// by segments. This function can be used to construct subsequent elements of
-// the constructed path. The return value contains the newly created file
+// MkDir creates a directory with a given name.
+//
+// The directory is represented with a file descriptor and a name (for
+// convenience). This function is meant to be used to construct subsequent
+// elements of some path. The return value contains the newly created file
```

Comment:
This comment is slightly confusing when reading the function declaration because it isn't clear if 'and a name' corresponds to `dir string` or `name string`. I suggest:
```
// The directory is represented with a file descriptor and its name (for
// convenience). This function is meant to be used to construct subsequent
// elements of some path. The return value contains the newly created file
// descriptor for the new directory or -1 on error.
func (sec *Secure) MkDir(dirFd int, dirFdName string, name string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) (int, error) {
...
```

That said, AFAICS, you aren't using `dir` (or as I suggested, `dirFdName`); perhaps just remove it?

Reply:
Oh we are using it for EROFS handling, unless the diff confused me.

In any case it is essential in a follow up patch that simplifies error handling (less stacked error messages) so I'd like to keep it.